Factor out pure balanceTransaction logic

Anviking opened the pull request:
### Motivation

Make property testing of the `balanceTransaction`  logic possible.

### Overview

- [x] Add `m` to `HasLogger` constraint rather than to assume IO
- [x] Make selectAssets run in `MonadRandom m` rather than IO
- [x] Let selectAssets take PParams directly, rather than fetching from NetworkLayer
- [x] Remove IO from assignScriptRedeemers
- [x] Factor out W.balanceTransaction running in `MonadRandom m`

### Issue Number

ADP-1182

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1272,139 @@ normalizeDelegationAddress s addr = do
+        -- transaction. A long-term fix is to handle this case properly during
+        -- balancing.
+    let transform s sel =
+            let (sel', _) = assignChangeAddresses @s undefined sel s
```
Oops. Will fix.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -297,6 +298,7 @@ import Cardano.Wallet.Primitive.AddressDiscovery.Shared
     )
 import Cardano.Wallet.Primitive.CoinSelection
     ( Selection
+    , SelectionCollateralRequirement (SelectionCollateralNotRequired, SelectionCollateralRequired)
```
Did you mean to do:
```suggestion
    , SelectionCollateralRequirement (..)
```

Or do you prefer to import these explicitly?

```suggestion
    , SelectionCollateralRequirement
        ( SelectionCollateralNotRequired
        , SelectionCollateralRequired
        )
```

Anviking replied:
Fixed, they ended up from me using LSP code actions to auto-import.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1272,139 @@ normalizeDelegationAddress s addr = do
+    tl
+    resolvedInputs
+    partialTx
+    = do
```
There are quite a lot of parameters. Would it be helpful to define a record type to hold them all?

This might also make it clearer at call sites which parameter is which (because each parameter is named). The errors when you accidentally omit a parameter can be more concise, too, rather than a long list of type errors for each parameter that has the wrong type.

Anviking replied:
I added the type
```
data PartialTx = PartialTx
    { sealedTx :: SealedTx
    , inputs :: [(TxIn, TxOut, Maybe (Hash "Datum"))]
    , redeemers :: [Redeemer]
    } deriving (Show, Generic, Eq)
```

(which I used in my prototype property test anyway), making the signature read

```
balanceTransaction
    :: forall m s k ctx.
        ( HasTransactionLayer k ctx
        , HasLogger m WalletWorkerLog ctx
        , GenChange s
        , MonadRandom m
        )
    => ctx
    -> ArgGenChange s
    -> (W.ProtocolParameters, Cardano.ProtocolParameters)
    -> TimeInterpreter (Either PastHorizonException)
    -> (UTxOIndex, Wallet s, Set Tx)
    -> PartialTx
    -> ExceptT ErrBalanceTx m SealedTx
```

Possibly there are more things to group, but I see nothing more as obvious as `PartialTx`, which I think conceptually makes sense.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1441,9 +1582,25 @@ selectAssets
+    tl = ctx ^. transactionLayer @k
+    tr = contramap MsgWallet $ ctx ^. logger
+
+-- | Less convenient and restrictive version of `selectAssets`.
```
I suspect Haddock only adds hyperlinks if you use "straight" single quotes.

```suggestion
-- | Less convenient and restrictive version of 'selectAssets'.
```

Anviking replied:
Mis-remembered; thanks!

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1441,9 +1582,25 @@ selectAssets
+    => Tracer m WalletLog
+    -> TransactionLayer k SealedTx
+    -> ProtocolParameters
+    -> SelectAssetsParams s result
```
Would it be helpful to add these other parameters into the `SelectAssetsParams` record type?

Anviking replied:
Diff outdated, marking as resolved.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1272,139 @@ normalizeDelegationAddress s addr = do
+    (pp, nodePParams)
+    (internalUtxoAvailable, wallet, pendingTxs)
+    tl
+    resolvedInputs
```
`externalInputs` would be a better name I think. They are resolved inputs yes, but only the external ones unknown of the wallet.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1441,9 +1582,25 @@ selectAssets
+    -> (s -> Selection -> result)
+    -> ExceptT ErrSelectAssets m result
+selectAssets' tr tl pp params transform = do
+    guardPendingWithdrawal
```
Why this function? What's the problem with pending withdrawal and this PR?

Anviking replied:
I didn't change anything,  just moved it from `selectAssets` to here.

KtorZ replied:
I know, but I don't get the rationale for creating `selectAssets'` and `selectAsset`. Why is this distinction needed?

Anviking replied:
Fixed. Keeping a single version of the function but with the following tweaks:

commit c76a655732f5abdabb00ec0faab00345deb73c9b
Author: Johannes Lund <johannes.lund@iohk.io>
Date:   Wed Oct 20 19:52:28 2021 +0200

    Let selectAssets take PParams directly, rather than fetching from NetworkLayer

    Passing the full NetworkLayer seems overkill, and would have required
    less elegant mocking to property test balanceTransaction
    (calling selectAssets).

commit 1c1d8abdd559156612ffde2330d8c739e0484596
Author: Johannes Lund <johannes.lund@iohk.io>
Date:   Wed Oct 20 19:47:11 2021 +0200

    Make selectAssets run in `MonadRandom m` rather than IO

commit 014c539ad765934f8cd01510d0ee0dc6dbc12971
Author: Johannes Lund <johannes.lund@iohk.io>
Date:   Wed Oct 20 19:44:48 2021 +0200

    Add `m` to `HasLogger` constraint rather than to assume IO

Anviking committed:
Make selectAssets run in `MonadRandom m` rather than IO

Anviking committed:
Let selectAssets take PParams directly, rather than fetching from NetworkLayer

Passing the full NetworkLayer seems overkill, and would have required
less elegant mocking to property test balanceTransaction
(calling selectAssets).

Anviking committed:
Remove IO from assignScriptRedeemers

Possible by adding the following function:
```
snapshot
    :: TimeInterpreter (ExceptT PastHorizonException IO)
    -> IO (TimeInterpreter (Either PastHorizonException))
```

Anviking committed:
Factor out W.balanceTransaction running in `MonadRandom m`

Which can be more easily property tested.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,193 @@ normalizeDelegationAddress s addr = do
                                   Transaction
 -------------------------------------------------------------------------------}
 
+data PartialTx = PartialTx
```
Should document this, I realised.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -549,6 +550,22 @@ expectAndThrowFailures = hoistTimeInterpreter (runExceptT >=> eitherToIO)
+-- | Pre-fetches a snapshot of the epoch history from the node, such that the
+-- resulting 'TimeInterpreter' doesn't require 'IO'.
+--
+-- Please concider /not/ using this function, as it disables all logging.
```
```suggestion
-- Please consider /not/ using this function, as it disables all logging.
```

Commit:
Regenerate nix

Anviking committed:
Add `m` to `HasLogger` constraint rather than to assume IO

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+                        SelectionCollateralNotRequired
+                } & padFeeEstimation partialTx
+
+        -- FIXME: The coin selection and reported fees will likely be wrong in
+        -- the presence of certificates (and deposits / refunds). An immediate
+        -- "fix" is to return a proper error from the handler when any key or
+        -- pool registration (resp. deregistration) certificate is found in the
+        -- transaction. A long-term fix is to handle this case properly during
+        -- balancing.
```
**Suggestion**: remove this comment.

**Justification**: this comment is now out of date, because coin selection already properly accounts for withdrawals, deposits and refunds.

Anviking replied:
No, it is still valid.

I'm sure coin-selection is self-consistent; it accounts for the [txDelegationAction](https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Transaction.hs#L253) just fine. However the real world is bigger. In `balanceTransaction` we have no guarantee to be able to map the raw `[ledgerCert]` to `Maybe DelegationAction`.

That's (at least one reason) why I think https://input-output-rnd.slack.com/archives/GBT05825V/p1634585642350800?thread_ts=1634549806.343500&cid=GBT05825V would make sense.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+                    , inputs
+                    , fst <$> (sel' ^. #collateral)
+                    , sel' ^. #change
+                )
```
```suggestion
                    , sel' ^. #change
                    )
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+    -- Once the coin-selection is done, we need to
+    --
+    -- (a) Add selected inputs, collateral and change outputs to the transaction
+    -- (b) Assign correct execution units to every redeemers
```
```suggestion
    -- (b) Assign correct execution units to every redeemer
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+    -- (d) Adjust fees and change output(s) to the new fees.
+    --
+    -- There's a strong assumption that modifying the fee value AND increasing
+    -- the coin value of change outputs does not modify transaction fees; or
```
```suggestion
    -- the coin values of change outputs does not modify transaction fees; or
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+    -- There's a strong assumption that modifying the fee value AND increasing
+    -- the coin value of change outputs does not modify transaction fees; or
+    -- more exactly, does not modify the execution units of scripts. This is in
+    -- principle a fair assumption because scripts validators ought to be
```
```suggestion
    -- principle a fair assumption because script validators ought to be
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+    -- the coin value of change outputs does not modify transaction fees; or
+    -- more exactly, does not modify the execution units of scripts. This is in
+    -- principle a fair assumption because scripts validators ought to be
+    -- unaware of change outputs. If their execution cost increases when change
```
```suggestion
    -- unaware of change outputs. If their execution costs increase when change
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+    -- more exactly, does not modify the execution units of scripts. This is in
+    -- principle a fair assumption because scripts validators ought to be
+    -- unaware of change outputs. If their execution cost increases when change
+    -- output increases, then it becomes impossible to guarantee that the fee
```
```suggestion
    -- output values increase, then it becomes impossible to guarantee that fee
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+    -- unaware of change outputs. If their execution cost increases when change
+    -- output increases, then it becomes impossible to guarantee that the fee
+    -- balancing will ever converge towards a fixed point. A script validator
+    -- doing such thing is considered bonkers and this is not a behavior we
```
```suggestion
    -- doing such a thing is considered bonkers and this is not a behavior we
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+    let candidateMinFee = fromMaybe (Coin 0) $
+            evaluateMinimumFee tl nodePParams candidateTx
+
+    let surplus = delta `Coin.distance` candidateMinFee
```
**Suggestion**: consider using `Coin.difference` here.

**Justification**: If this code expects that `delta >= candidateMinFee`, then `Coin.distance` will give a nonsensical answer if this expectation is violated. It's probably better to compute a `surplus` of `0` in this case.

Anviking replied:
Sounds very sane 👍

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+            (\(_,o) -> (o, Nothing)) <$> L.find (\(i',_) -> i == i') (extraInputs update)
+
+    extractFromTx tx =
+        let (Tx _id _fee _coll _inps outs wdrlMap meta _vldt, toMint, toBurn) = decodeTx tl tx
```
**Suggestion**: Consider using `NamedFieldPuns` here, or otherwise use label accessors (`view #outputs`).

**Justification**: This code only cares about a subset of the record fields. Pattern matching on everything also makes the line much longer than our line length limit.

Anviking replied:
Possibly, but I think there might be usefulness to showing which fields are ignored. In-doubt, I'd prefer to keep it as is for the time being. I will move `decodeTx` to a new line, however.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+            wdrl = WithdrawalSelf
+                (error $ "WithdrawalSelf: reward-account should never been use "
+                      <> "when balancing transactions but it was!"
+                )
```
Suggestion: remove variable alignment and fix the broken error message.

```suggestion
                (error $ unwords
                    [ "WithdrawalSelf: reward account should never have been used"
                    , "when balancing a transaction, but it was!"
                    ]
                )
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+                )
+                (error $ "WithdrawalSelf: derivation path should never been use "
+                      <> "when balancing transactions but it was!"
+                )
```
Suggestion: remove variable alignment and fix the broken error message.

```suggestion
                (error $ unwords
                    [ "WithdrawalSelf: derivation path should never have been used"
                    , "when balancing a transaction, but it was!"
                    ]
                )
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+    -- `evaluateMinimumFee`.
+    --
+    -- TODO: This logic needs to be consistent with how we call `selectAssets`,
+    -- so it would be good to join them into some single helper.
```
Suggestion: use Haddock-friendly quotation marks.

```suggestion
    -- 'balanceTransaction'. As a workaround we add some padding using
    -- 'evaluateMinimumFee'.
    --
    -- TODO: This logic needs to be consistent with how we call 'selectAssets',
    -- so it would be good to join them into some single helper.
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1264,6 +1284,199 @@ normalizeDelegationAddress s addr = do
+            extraMargin = Coin $ ceiling $ (*) b $ fromIntegral
+                $ sizeOfScriptIntegrityHash
+                + sum ((+sizeOfRedeemerCommon) . BS.length . redeemerData <$> redeemers)
+              where
```
Suggestion: use `F.foldMap` and extract out a function `sizeOfRedeemer`.

```suggestion
                + F.foldMap sizeOfReedemer redeemers
              where
                sizeOfRedeeder
                    = (+ sizeOfRedeemerCommon) . BS.length . redeemerData
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2517,14 +2347,15 @@ delegationFee ctx (ApiT wid) = do
+        pp <- liftIO $ NW.currentProtocolParameters (wrk ^. networkLayer)
         deposit <- W.calcMinimumDeposit @_ @s @k wrk wid
-        mkApiFee (Just deposit) [] <$> W.estimateFee (runSelection wrk deposit w)
+        mkApiFee (Just deposit) [] <$> W.estimateFee (runSelection wrk pp deposit w)
```
Suggestion: Wrap line to fit within line limit.

```suggestion
        mkApiFee (Just deposit) [] <$>
            W.estimateFee (runSelection wrk pp deposit w)
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -549,6 +550,22 @@ expectAndThrowFailures = hoistTimeInterpreter (runExceptT >=> eitherToIO)
+-- | Pre-fetches a snapshot of the epoch history from the node, such that the
+-- resulting 'TimeInterpreter' doesn't require 'IO'.
+--
+-- Please consider /not/ using this function, as it disables all logging.
```
👍🏻

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -932,15 +931,14 @@ _assignScriptRedeemers (toAlonzoPParams -> pparams) ti resolveInput redeemers tx
-        -> ExceptT ErrAssignRedeemers m (Map Alonzo.RdmrPtr (Either ErrAssignRedeemers Alonzo.ExUnits))
-    evaluateExecutionUnits indexedRedeemers alonzoTx = withExceptT ErrAssignRedeemersPastHorizon $ do
+        -> Either ErrAssignRedeemers (Map Alonzo.RdmrPtr (Either ErrAssignRedeemers Alonzo.ExUnits))
+    evaluateExecutionUnits indexedRedeemers alonzoTx = left ErrAssignRedeemersPastHorizon $ do
```
Request: fit these lines within the line length limit.

```suggestion
        -> Either ErrAssignRedeemers
            (Map Alonzo.RdmrPtr (Either ErrAssignRedeemers Alonzo.ExUnits))
    evaluateExecutionUnits indexedRedeemers alonzoTx =
        left ErrAssignRedeemersPastHorizon $ do
```

jonathanknowles reviewed (commented):
Hi @Anviking 

This PR seems like a very positive step towards making `balanceTransaction` more testable. Thank-you for making it! 👍🏻 

I only have a few suggestions and requests: some of these apply to code that was moved (and therefore not "changed" by this PR per se), but I think it would still be good to clean up the highlighted areas before merging.

Anviking committed:
Review suggestions

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

Anviking committed:
Use Coin.difference instead of Coin.distance

In response to:
https://github.com/input-output-hk/cardano-wallet/pull/2970#discussion_r737107680

Anviking committed:
Fixup: wrap lines

jonathanknowles reviewed (approved):
Hi @Anviking 

Many thanks again for making this PR. It looks good to me!
